ably/options.go

```diff
@@ -224,11 +232,10 @@ type ClientOptions struct {
 	Trace *httptrace.ClientTrace
 }
 
-func NewClientOptions(key string) *ClientOptions {
+func NewClientOptions() *ClientOptions {
```

Comment:
Given my lack of Go knowledge this is just an observation, but is the **only** way to enforce the specified default value for these two members via this constructor? Reading [this answer](https://stackoverflow.com/a/37521991) it seems to me that we would need quite a big refactor to control and encapsulate this better. Is that worth doing? Or is this just an unavoidable compromise in this language?

Reply:
@QuintinWillison Yes, this is basically the best we can do without getting too weird. It doesn't matter that much in this case since instantiating the library is supposed to run a number of validations on the passed ClientOptions anyway.

Reply:
Another convention is not having an exposed `Options` struct but have it be internal and use functional options as explained [here](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis) (which gives the library full control over things like default values since the caller can't just do something like `NewClient(&ClientOptions{})` rather than using the constructor).

Reply:
I would support this, but it does go against our preference for similarity across the APIs on different platforms.

Reply:
Something akin to functional options do seem a good idea here. I think it's not so much a deviation from the spec as it's a particular way to implement it.

How about this?

```go
type ClientOptions []ClientOption

type ClientOption func(*clientOptions)

func (o ClientOptions) Key(key string) ClientOptions {
    return append(o, func(o *clientOptions) {
        o.Key = key
    })
}

options := ably.ClientOptions{}.
    Key("...").
    AuthCallback(auth).
    AutoConnect(false)

ably.NewRealtime(options...)

// And we could maybe provide top-level shortcuts to match the constructor overloads defined in the spec:   
ably.NewRealtime(ably.Key("..."))
ably.NewRealtime(ably.Token("..."))
```

I wouldn't say this deviates from the spec, provided the translation from the abstract IDL to Go is consistent with naming and types.

Reply:
I support doing this ^. I know it's a breaking change, but it's a better API. Just allowing clients to create the zero value `ClientOptions{}` is clearly not going to work, and if you're going to force the use of a `NewClientOptions()` factory method, you might as well do it properly.

Reply:
@tcard I would probably only export the option function type and the option methods to lower the surface area of the API, for example:

```go
type clientOptions struct {
	key string
}

type ClientOption func(*clientOptions)

func Key(key string) ClientOption {
	return func(opts *clientOptions) {
		opts.key = key
	}
}

func NewClient(opts ...ClientOption) *Client {
	options := newClientOptions(opts...)
}

client := ably.NewClient(ably.Key("..."))
```

Reply:
@lmars I'd swear I responded to this at the time, but apparently I didn't (or something ate it).

The problem with that approach is that we have _a lot_ of options, so putting them all in the same top-level namespace is problematic for discoverability and potential name conflicts.

My proposal above has been implemented in #145 and #146 (not that we can't still change it).

Reply:
Ah, I see now that I replied [in the pull request](https://github.com/ably/ably-go/pull/145#issuecomment-610958179).